src/libraries/System.Security.Cryptography.Algorithms/src/System/Security/Cryptography/HKDF.cs

```diff
@@ -134,35 +134,66 @@ private static void Expand(HashAlgorithmName hashAlgorithmName, int hashLength,
-                    hmac.AppendData(info);
-                    counter = (byte)i;
-                    hmac.AppendData(counterSpan);
+                tempInfoBuffer = info.Length > MaxStackInfoBuffer ?
+                    (rentedTempInfoBuffer = CryptoPool.Rent(info.Length)).AsSpan(0, info.Length) :
+                    (stackalloc byte[MaxStackInfoBuffer]).Slice(0, info.Length);
 
-                    if (remainingOutput.Length >= hashLength)
-                    {
-                        t = remainingOutput.Slice(0, hashLength);
-                        remainingOutput = remainingOutput.Slice(hashLength);
-                        GetHashAndReset(hmac, t);
-                    }
-                    else
+                info.CopyTo(tempInfoBuffer);
+                infoBuffer = tempInfoBuffer;
+            }
+            else
+            {
+                infoBuffer = info;
+            }
+
+            try
+            {
+                using (IncrementalHash hmac = IncrementalHash.CreateHMAC(hashAlgorithmName, prk))
+                {
+                    for (int i = 1; ; i++)
                     {
-                        if (remainingOutput.Length > 0)
+                        hmac.AppendData(t);
+                        hmac.AppendData(infoBuffer);
+                        counter = (byte)i;
+                        hmac.AppendData(counterSpan);
+
+                        if (remainingOutput.Length >= hashLength)
                         {
-                            Debug.Assert(hashLength <= 512 / 8, "hashLength is larger than expected, consider increasing this value or using regular allocation");
-                            Span<byte> lastChunk = stackalloc byte[hashLength];
-                            GetHashAndReset(hmac, lastChunk);
-                            lastChunk.Slice(0, remainingOutput.Length).CopyTo(remainingOutput);
+                            t = remainingOutput.Slice(0, hashLength);
+                            remainingOutput = remainingOutput.Slice(hashLength);
+                            GetHashAndReset(hmac, t);
+                        }
+                        else
+                        {
+                            if (remainingOutput.Length > 0)
+                            {
+                                Debug.Assert(hashLength <= 512 / 8, "hashLength is larger than expected, consider increasing this value or using regular allocation");
+                                Span<byte> lastChunk = stackalloc byte[hashLength];
+                                GetHashAndReset(hmac, lastChunk);
+                                lastChunk.Slice(0, remainingOutput.Length).CopyTo(remainingOutput);
+                            }
+
+                            break;
                         }
-
-                        break;
                     }
                 }
             }
+            finally
+            {
+                CryptographicOperations.ZeroMemory(tempInfoBuffer);
```

Comment:
isn't cryptopool already zeroing bytes?

Reply:
The `CryptoPool` `Return` below does not do the zero because `tempInfoBuffer` might be stack data, or it might be pool data. I erred on the size of zeroing the data regardless of which since we don't know what developers might be feeding in to the `info` parameter. Since we have the `ZeroMemory`, we don't need to clear it again when returning to the pool.